syntax = "proto3";

service PictureService {
    rpc Upload(UploadRequest) returns (UploadResponse);
    rpc Search(SearchRequest) returns (SearchResponse);
    rpc Download(DownloadRequest) returns (DownloadResponse);
    rpc Delete(DeleteRequest) returns (DeleteResponse);
    rpc Like(LikeRequest) returns (LikeResponse);
    rpc List(ListRequest) returns (ListResponse);
    rpc Health(HealthRequest) returns (HealthResponse);


    // --- 2PC EXTENSIONS --John Song-//
    // Phase 1: Voting
    rpc VoteRequest(VoteArgs) returns (VoteReply);
    // Phase 2: Decision
    rpc GlobalCommit(DecisionArgs) returns (Ack);
    rpc GlobalAbort(DecisionArgs) returns (Ack);
    //------------------------------//

}

message UploadRequest {
    string filename = 1;
    bytes data = 2;
}

message UploadResponse {
    bool success = 1;
    string node = 2;
}

message SearchRequest {
    string filename = 1;
}

message SearchResponse {
    bool found = 1;
    string node = 2;
    int32 likes = 3;
}

message DownloadRequest {
    string filename = 1;
}

message DownloadResponse {
    bytes data = 1;
    bool found = 2;
}

message DeleteRequest {
    string filename = 1;
}

message DeleteResponse {
    bool success = 1;
}

message LikeRequest {
    string filename = 1;
}

message LikeResponse {
    bool success = 1;
    int32 likes = 2;
}

message ListRequest {}

message ListResponse {
    map<string, PictureMetadata> pictures = 1;
}

message PictureMetadata {
    int32 likes = 1;
    string node = 2;
}

message HealthRequest {}

message HealthResponse {
    string status = 1;
    string node = 2;
}

// --- 2PC MESSAGES ---
message VoteArgs {
    string transaction_id = 1;
    string coordinator_id = 2;
    string filename = 3;
}

message VoteReply {
    bool vote_granted = 1;
    string node_id = 2;
}

message DecisionArgs {
    string transaction_id = 1;
    string coordinator_id = 2;
    string filename = 3;
}

message Ack {
    bool success = 1;
    string node_id = 2;
}